plutus-ir/test/ParserSpec.hs

```diff
@@ -0,0 +1,119 @@
+{-# LANGUAGE FlexibleContexts #-}
+module ParserSpec (parsing) where
+
+import           PlutusPrelude
+
+import           Common
+import           TestLib
+
+import           Control.Monad
+import           Control.Monad.Except
+import           Control.Monad.State
+
+import qualified Data.ByteString.Lazy         as BSL
+
+import           Language.PlutusIR
+import qualified Language.PlutusIR.Parser     as Parser
+
+import qualified Language.PlutusCore          as PLC
+import qualified Language.PlutusCore.Constant as PLC
+import qualified Language.PlutusCore.Pretty   as PLC
+
+import           Hedgehog                     hiding (Var)
+import qualified Hedgehog.Gen                 as Gen
+import qualified Hedgehog.Range               as Range
+
+import           Text.Megaparsec              hiding (failure)
+
+import           Test.Tasty
+import           Test.Tasty.Hedgehog
+
+propParser :: Property
+propParser = property $ do
+    prog <- forAll genProgram
+    let pretty = prettyString prog
+    ast <- case Parser.parse "test" Parser.program pretty of
+               Left e -> do
+                   annotate pretty
+                   annotate $ parseErrorPretty e
+                   failure
+               Right p -> pure p
+    pretty === prettyString ast
+
+parsing :: TestNested
+parsing = return $ testGroup "parsing"
+    [testProperty "parser round-trip" propParser]
+
+
+genTyName :: MonadGen m => m (TyName ())
+genTyName = TyName <$> genName
+
+-- TODO make this robust against generating identfiers such as "fix"?
+genName :: MonadGen m => m (Name ())
+genName = Name () <$> name' <*> int'
+    where int' = PLC.Unique <$> Gen.int (Range.linear 0 3000)
+          name' = Gen.text (Range.linear 1 20) Gen.lower
+
+simpleRecursive :: MonadGen m => [m a] -> [m a] -> m a
+simpleRecursive = Gen.recursive Gen.choice
+
+genKind :: MonadGen m => m (Kind ())
+genKind = simpleRecursive nonRecursive recursive
+    where nonRecursive = pure <$> sequence [Type, PLC.Size] ()
+          recursive = [KindArrow () <$> genKind <*> genKind]
+
+genBuiltinName :: MonadGen m => m PLC.BuiltinName
+genBuiltinName = Gen.element PLC.allBuiltinNames
+
+genBuiltin :: MonadGen m => m (PLC.Builtin ())
+genBuiltin = PLC.BuiltinName () <$> genBuiltinName
+
+genConstant :: MonadGen m => m (PLC.Constant ())
```

Comment:
Shouldn't we be able to reuse a bunch of the generation stuff from the PLC lib?

Reply:
Yes, but it's not exported, as it's actually all in the test suite - maybe they should be moved into the library, but it feels odd exposing something so internal

Reply:
I don't think that exposing generators for testing is that weird, it's quite useful actually. `language-plutus-core` already exports a bunch of generator stuff.

In the glorious future of multiple public libraries per package we'd put them in a separate public lib, but for now I say just expose them.